[all] Update containers to support c++x11 features

matthieu-nesme opened the pull request:
solving #107 

Adding move constructor and move assignment operator to helper containers.

matthieu-nesme committed:
Replacing helper::set with stream operator implementations of std::set.
So we can truly benefit of stl improvements and it is easy to create a Data<set>.

matthieu-nesme committed:
minor cleaning

matthieu-nesme committed:
Replacing helper::list with stream operator implementations of std::list.
So we can truly benefit of stl improvements and it is easy to create a Data<list>.

matthieu-nesme committed:
Replacing helper::deque with stream operator implementations of std::deque.
So we can truly benefit of stl improvements and it is easy to create a Data<deque>.

matthieu-nesme committed:
Directly aliasing boost::container::stable_vector as sofa::helper::stable_vector
so we can truly benefit of boost improvements.

matthieu-nesme committed:
adding std::pair stream operator implementations

matthieu-nesme committed:
fixing compilation

matthieu-nesme committed:
Adding move constructor and move assignment operator to helper::vector and helper::SVector.
(I hope it is ok, I am not yet familiar with that c++11 stuff)

matthieu-nesme committed:
fixing compilation

guparan commented on SofaKernel/framework/sofa/helper/set.h:
```diff
@@ -2,18 +2,18 @@
 *                (c) 2006-2016 INRIA, USTL, UJF, CNRS, MGH                    *
 *                                                                             *
-* This library is free software; you can redistribute it and/or modify it     *
+* _set library is free software; you can redistribute it and/or modify it     *
```
I guess some "find+replace" went wrong here.

guparan replied:
You forgot this one ;)

matthieu-nesme committed:
minor comment

matthieu-nesme commented:
:D corrected

guparan commented on SofaKernel/framework/sofa/helper/set.h:
```diff
@@ -13,7 +13,7 @@
 *                                                                             *
 * You should have received a copy of the GNU Lesser General Public License    *
-* along with this library; if not, write to the Free Software Foundation,     *
+* along with _set library; if not, write to the Free Software Foundation,     *
```
And this one ;)

guparan reviewed (changes requested):
Thanks for your fix but you forgot two "_set" typos :(

matthieu-nesme committed:
minor comment

matthieu-nesme commented:
Corrected again...
This code is pure c++, no need for special assignees.
I let you update your changelog if you want to incorporate this PR.

damienmarchal committed:
Update CHANGELOG.md

damienmarchal committed:
Merge branch 'master' into containers

damienmarchal committed:
Merge branch 'master' into containers

damienmarchal commented:
@guparan  Is there is still something to change ? 
As this does not change the API or the scene it can be merged for the 16.12

JeremieA commented:
Hello,

The issue I have with this approach is that this make it very invasive, as Sofa is adding operators in std namespace for all types, even not its own. So within an application that could also does the same, we would get conflicts / redefinition / ambiguity compiler errors, or worse silently override the application's own operators that could have relied on a different string syntax.

I agree that the approach of subclassing the standard containers just to be able to package them within a Data is wrong. In our current branch we are refactoring the approach, relying more on `DataTypeInfo` to generically access features like string parsing, without having to put them in the type itself. But this is work in progress so it is not ready to be pushed yet. So in the meantime this change could be pushed, but with the understanding that it may be a temporary solution only.

Jeremie.

matthieu-nesme commented:
@JeremieA I was waiting for this remark :) I totally agree. My point was to wait for an eventual (quite improbable) conflict. But if you think the conflict could be silent, indeed it's a problem.

damienmarchal commented:
This PR: 
  - [ ] successfully compile on each configuration on the CI.
  - [x] does not generates new warnings or test failure. 
  - [x] does not break existing scenes.
  - [x] does not break API compatibility.  
  - [x] is now 1 week old.
  - [x] has been reviewed and agreed to be transitional.

damienmarchal committed:
Merge branch 'master' into containers

Commit:
ADD \todo comments to keep track of the temporary changed introduced in PR#113

Comment is:
/// adding string serialization to std::deque to make it compatible with Data
/// \todo: refactoring of the containers required
/// More info PR #113: https://github.com/sofa-framework/sofa/pull/113

damienmarchal commented:
Hi all,

I have opened a new PR in WIP to handle the cleaning of set & vector.

It currently contains an attempt to remove the std namespace "pollution" in commit https://github.com/sofa-framework/sofa/pull/278/commits/d3ecab1eaa7f2f29242aacf646910b2eb3ab47dd

Feedback & contribution in #278
